### PR TITLE
MAINT: Overhaul and refactor NPY_OS_WIN*

### PR DESCRIPTION
**Pull request by cbrt64**
When #20884 added `NPY_OS_WIN64`, the meaning of `NPY_OS_WIN32` became ambiguous; some might take it to mean "32-bit only", while others might assume its original meaning of "any Windows".

Besides that, the directives as written confuse the nature of `_WIN32`, which I understand to be always defined in any native Windows toolchain. Thus `#define NPY_OS_WIN64` would never be reached.

This commit disambiguates by removing `NPY_OS_WIN32` (in case anyone uses it in future without looking again at its definition), and adding `NPY_OS_WINDOWS` with `NPY_OS_WIN_xxBIT` children, in the pattern of `NPY_OS_BSD`.

It also removes `NPY_OS_MINGW`, which for the same reason never got defined. MINGW is not an OS, but a toolchain like MSVC. Maybe there could be `NPY_BUILD_MINGW` and `NPY_BUILD_MSVC` or some such?

I'm not attached to this particular solution, or the names I've chosen. Input is welcome.

@Mousius: I won't pretend to understand the stack alignment problem in #23399. In [loops_trigonometric.dispatch.c.src](https://github.com/Mousius/numpy/commit/fe5472fa4eae131ff9646d7c980c6c4081c10386#diff-4ce69850e0011abbbeb8d176091ca7297e13cc85da738df8c5d0616f54a3365aR47), can you confirm you wanted to force inlining on all windows systems, not just 32-bit?

### CODE REVIEW

**Comment by Mousius** [comment]
@cbrt64 iirc the stack alignment is helping to encourage the compiler to place things in the correct registers whilst also working around the compiler not wanting to index into them directly (we don't have `get`/`set` for lanes iirc). The `FINLINE` was for the 32-bit windows calling pattern which places everything on stack, the theory was it was overflowing due to some alignment issues; on 64-bit windows I believe this function should be safe as the 64-bit calling convention will use the floating point registers instead.

TLDR; It **should** only need to inline on 32-bit windows

**Comment by seberg** [comment]
So, the current main branch is broken?  I am OK with changing the macros a bit, but it should have a release note.  It is probably also too late for doing this for the next release.

@cbrt64 can I convince you to make a minimal fixup to merge quickly?  Sorry that I messed it up on the other PR.

**Comment by cbrt64** [comment]
I can do that. I'm thinking just use `_WIN64` like downstream is the simplest for now. Also it's already tested. Do you want a new PR, and keep this one for later, or just force push here?

**Comment by charris** [comment]
Going to push this off to 2.2, but it looks like is needs a refresh in any case. Maybe close?

**Comment by charris** [comment]
Pushing off to 2.3. Is this still relevant?
